pulsar-common/src/test/java/org/apache/pulsar/common/validator/ConfigValidationTest.java

```diff
@@ -0,0 +1,148 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.common.validator;
+
+import org.testng.annotations.Test;
+
+import java.util.*;
+import static org.testng.Assert.*;
+
+public class ConfigValidationTest {
+
+    private final List<String> testStringList = Arrays.asList(new String[]{"foo", "bar"});
+    private final List<Integer> testIntegerList = Arrays.asList(new Integer[]{0, 1});
+    private final Map<String, Integer> testStringIntegerMap = new HashMap<String, Integer>() {
+        {
+            put("one", 1);
+            put("two", 2);
+        }
+    };
+    private final Map<String, String> testStringStringMap = new HashMap<String, String>() {
+        {
+            put("one", "one");
+            put("two", "two");
+        }
+    };
+    private final String topic = "persistent://public/default/topic";
+
+    public static class TestValidator extends Validator {
+        @Override
+        public void validateField(String name, Object o) {
+            if (o instanceof String) {
+                String value = (String)o;
+                if (!value.startsWith("ABCDE")) {
+                    throw new IllegalArgumentException(String.format("Field %s does not start with ABCDE", name));
+                }
+            } else {
+                throw new IllegalArgumentException(String.format("Field %s is not a string", name));
+            }
+        }
+    }
+
+    class TestConfig {
```

Comment:
I like the idea of having type checking. The naming is a bit confusing to me something like makes more sense to me. 
```
class TestConfig {
       @NotNull
        public String stringValue;
        
       @PositiveNumber or @Number(positive=true)
        public Integer positiveNumber;
        
       @List(itemType = Integer.class) or @ListType(itemType = Integer.class) or @IntegerList
        public List integerList;
        
       @Map(keyType = String.class, valueType = Integer.class) or @MapType(keyType = String.class, valueType = Integer.class)
        public Map stringIntegerMap;
        
       @List(itemType = String.class) or @StringList
        public List stringList;
        
       @TopicName
        public String topic;
        
       @CustomType(validatorClass = TestValidator.class)
        public String customString;
    }
```
I think annotations should describe what the variable type is as opposed to describing how to test it. The testing is implied. 

Another thing to think about is how to handle fields that should be required to have a value.

Thoughts?

Thoughts? Al

Reply:
I don't fee strongly about it either way.   However, if we feel we can simplify the naming of annotations and still be intuitive to the end user, we should.  The current annotations are borrowed from Storm that I created a while ago:

https://github.com/apache/storm/blob/master/storm-client/src/jvm/org/apache/storm/validation/ConfigValidationAnnotations.java

Others have name annotations in a more simplified way:

https://www.baeldung.com/javax-validation

Reply:
I have changed the annotations. wrt fields that should have a value, isn't that covered by NotNull?

Reply:
Can NotNull be applied in combination with another annotation like this? Or does an annotation like `@TopicName` or `@List` imply not null?
```
@NotNull
@TopicName
 public String topic;
```

Reply:
Multiple annotations are possible on a field.
Thus
@NotNull
@TopicName implies that it should be non null and is a topic. However I believe that TopicName already ensures that its not null

Reply:
Sounds good and LGTM +1